### PR TITLE
Deprecate mapped outputs

### PR DESCRIPTION
**Pull request by wolfs**
Instead of using
```groovy
@OutputFiles Map<String, File> reportOutputs
```
use
```groovy
@Nested Map<String, Report> reports
...

class Report {
    @OutputFile File reportLocation
}
```

### CODE REVIEW

**Comment by melix** [comment]
What is the incentive to do this? Out of nowhere, it just feels like an arbitrary complication of something that used to be simple to do.

**Comment by wolfs** [comment]
@melix We introduced the mapped outputs version of outputs as a quick workaround when the AGP plugin required this kind of feature. Later, we started supporting nested polymorphic inputs for `Map`s and we think that is what should be used if a dynamic number of inputs or outputs is required.

If we continue supporting `Map<String, File>` for outputs, then we probably should also support `Map<String, File>` for input files (`@InputFiles`/`@InputDirectories`).

It seems like the use case is not very frequent, and the main ones we saw are better solved using `@Nested`, for example for the different kinds of reports of code quality tasks.

Note that you can use a simple `Iterable<File>` as `@OutputFiles`, which is enough for most use cases, so no `Map` required.

Do you have a compelling use-case for using mapped output files?

File: subprojects/core/src/integTest/groovy/org/gradle/api/internal/project/taskfactory/TaskPropertyNamingIntegrationTest.groovy [hunk 1]
```diff
@@ -100,6 +100,7 @@ class TaskPropertyNamingIntegrationTest extends AbstractIntegrationSpec {
             }
         """
         when:
+        executer.expectDeprecationWarnings(2)
```
Comment (lptr) [inline subprojects/core/src/integTest/groovy/org/gradle/api/internal/project/taskfactory/TaskPropertyNamingIntegrationTest.groovy]: Should we add a `@Nested Map` property to this test?

File: subprojects/docs/src/docs/userguide/upgrading_version_5.adoc [hunk 2]
```diff
@@ -80,6 +80,41 @@ _and_ that Ivy dependency was substituted (using a `resolutionStrategy.force`, `
...
+public interface Report {}
+
+public class DirectoryReport {
+    @Input
+    String getName() { /* ... */ }
+    @OutputDirectory
+    File getOutputDirectory() { /* ... */ }
+}
+----
+
```
Comment (lptr) [inline subprojects/docs/src/docs/userguide/upgrading_version_5.adoc]: Nice. We should mention the deprecation in `notes.md`, too.

**Comment by melix** [comment]
I don't have a specific use case, but I just recognize that having to create a `Report` class as one level of indirection to do this seems redundant. I understand we support `Iterable<File>` but in this case it means we should also deprecate and require `@Nested Iterable<Report>`. I'd be happier if we special cased types we know about.

File: subprojects/core/src/integTest/groovy/org/gradle/api/internal/project/taskfactory/TaskPropertyNamingIntegrationTest.groovy (hunk 1 above)
Comment (wolfs) [reply subprojects/core/src/integTest/groovy/org/gradle/api/internal/project/taskfactory/TaskPropertyNamingIntegrationTest.groovy]: The tests for nested properties are in `NestedInputIntegrationTest`, in particular the one for the map is [here](https://github.com/gradle/gradle/blob/9f9a664e4e2ae2539849bc10b4ce4e2e9dbcbbbc/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/NestedInputIntegrationTest.groovy#L715).

I would currently leave them there.

File: subprojects/docs/src/docs/userguide/upgrading_version_5.adoc (hunk 2 above)
Comment (wolfs) [reply subprojects/docs/src/docs/userguide/upgrading_version_5.adoc]: I mentioned it in the release notes as well.

**Comment by netvl** [comment]
I'd like to add as a user (by coincidence I've encountered exactly this issue yesterday - see #8646), that it makes no sense to me why `Iterable<File>` are supported and `Map<String, File>` are not and  require a level of indirection. There seem to be no conceptual difference between these two cases, because in both of them we have a collection of `Files` (and you can get `Iterable<File>` from `Map<?, File>` just by calling the `values()` method).

In my particular use case (referenced in that ticket) the map of files *is* the semantic input of the task, and the necessary wrapper would have no semantic meaning at all, so my map would look like `MapProperty<String, FileWrapper>` since there is no better name for the value type. This looks really ugly and feels like a hack.

But I do agree that this kind of distinction between inputs and outputs is even worse. Either both input and output properties should not allow maps with files (and this should be pointed out in documentation), or both should be allowed.

**Comment by netvl** [comment]
Oh, and by the way - for my use case this would require introducing *two* wrappers for files, because in one task I have this `Map<String, File>` as `@OutputFiles`, and in another task I have them as `@InputFiles`. Therefore, to connect these two tasks I would also need to apply a conversion method. This looks like a really great pain without any reason.

**Comment by wolfs** [comment]
We decided not to deprecate mapped output files, but support [mapped input files](#8646) as well.

**Comment by melix** [comment]
I just want to mention, for completeness, that there is one advantage with your technique @wolfs, which is that the nested object can mix inputs and outputs:

```
@Nested
Map<String, Whatever> config

class Whatever {
    @InputFile File inputFile
    @OutputFile File outputFile
}
```

```

**Pull request closed without merging**